### DISCUSSION TITLE
ファイル一覧の日付欄の表示で秒単位の値が「00」になる件

### DISCUSSION THREAD

**Discussion by Power-Engine** [ffftp/ffftp#288] [Q&A] [answered]
お世話になります

タイトルのとおり、FFFTPのファイル一覧に表示されるタイムスタンプの秒単位の値が、一律ゼロになる問題を確認しました。
特に《**ローカルのファイルの秒単位がゼロ**》になる点が、本件提起対象（問題点）です。
なおリモートのファイルも秒単位がゼロになりますが、それは当該FTPサーバーから（MLSDでなく）LISTコマンドで取得された結果と理解できます（∵時分まで取得）。

つきましては考えられる原因・対策の有無につきまして、貴チーム見解を賜りたく、詳細を下記致します。
先ずはご確認いただけると幸いです。

　　　　記

〔影響状況〕
現在まで把握している限り、この問題で生じている影響は、不明です。
想定される影響として、ファイル一覧上の表示だけならともかく、ファイル・タイムスタンプの実態と異なる表示値異常が、ホストへのアップロードやホスト～ローカル間の同期処理に影響するのではないか、と危惧しています。

〔再現性（頻度、再現条件）〕
現在この問題事象は100%発生します。
かつて（同じシステムでも）この問題は発生していなかったと認識しています。
何らか変化があったと思われますが発生条件（再現条件）は不明です。
FFFTPを最新版に（4.7→5.1）しても、同じ問題が生じることを確認しました。

以上

**Comment by sayurin** [comment]
ffftpでは公開当初より内部データというものがほとんどなく、画面表示がすべてで、表示されていない情報・非表示にした情報は存在していないという作りになっています。
信じられないかもしれませんが、ファイル一覧を画面に表示したらその情報は捨てられます。アップロード等を行う際には画面に表示している文字列を解析して時刻情報を復元しています。そのため、秒を表示していれば秒も復元されますが、表示していなければ秒は復元できず 0 として扱われます。

現在、内部処理を改善していて、いちいち文字列解析せず内部情報を保持するように改良中です。その前段として、内部情報に必ず出現する文字列の扱いを改善が最近完了したところです…。

**Reply by Power-Engine** [reply]
早速のご返答ありがとうございました。

＞ファイル一覧を画面に表示したらその情報は捨てられます。アップロード等を行う際には画面に表示している文字列を解析して時刻情報を復元しています。

１、確かに意外な設計ですが、これは私の過日の[投稿](https://github.com/ffftp/ffftp/discussions/276#discussioncomment-773576)に対する回答、特に次の問題提起に対し「原因」の回答を賜れたもの、と感謝申し上げます。：
＞真逆な現象が見られました。：
＞
＞　タイムスタンプは旧いのに内容が新しい（更新されている）
＞
＞この現象は「見かけのディレクトリ情報」がダウンロード結果に反映される結果であるため、適宜手動の
＞［表示を更新］ON操作を実行して最新のディレクトリ情報にすることで回避可、と理解します。
＞ここで「見かけのディレクトリ情報」とは、FFFTPの使用経験で得た次のような知見に基づく表現です。：
＞　FFFTPの「ホスト側のファイル一覧」では、「ローカル側のファイル一覧」と異なり、
＞　ホスト側でファイルが上書き更新されても、タイムスタンプは自動的に更新されない。

つまり、FFFTP利用上で次のような注意事項があるものと理解致します。：
　ファイル一覧に表示される時刻情報は、ホスト側（リモート）もローカル側も「表示どおり」のものがすべてであり、ファイル転送処理等においては最新情報に更新（リフレッシュ）が自動的にされることなく、そのまま使用される。

２、関連する開発状況のご教示、感謝申し上げます。：
＞文字列解析せず内部情報を保持するように改良中

この「表示どおり」を「内部に保持」は、解析時間不要になり高速化の改良かと存じます。
しかし、依然としてファイル転送処理等において「最新情報に更新が自動的にされることなし」のようで、少なくとも過日の問題提起（上記「タイムスタンプは旧いのに内容が新しい（更新されている）」事象）への「対策」の回答ではないように見えます。

３、ただ、本件提起対象（問題点）《**ローカルのファイルの秒単位がゼロ**》へ回答は未、と認識致します。

そもそも、ローカルのファイル一覧の元になるディレクトリ情報から秒単位まで読み取って表示する、と考えます。
その結果であるはずのファイル一覧に表示されるタイムスタンプの秒単位の値が、一律ゼロになる、という問題です。
この問題事象の当方環境下での発生状況は、初めて発見してから現在に至るまで、一度も回復せず、100%です。
「表示を更新」しても変わらず。バージョンを最新（5.1）にしても変わらず。

この事象について考えられる原因・対策の有無につきまして、貴チーム見解を賜りたいです。


以上／ご協力に感謝申し上げます。

**Reply by sayurin** [reply]
もしかして環境設定に「ファイルのタイムスタンプの秒単位を表示する」という項目が存在することをご存じないということでしょうか？
表示すれば秒を扱いますし、非表示では先述の通り存在しないものとして扱われ常に`0`になります。

**Reply by Power-Engine** [reply]
ご指摘ありがとうございます。
その設定項目をOnとせずにはファイル一覧に秒単位を表示できないことは**元より**承知です。
実際、本件オープン時の投稿では割愛しましたが、ローカル側もリモート側も、まさにその環境設定で表示させています。
その設定項目をOnとしたにもかかわらず変秒単位の表示を「しない」のではなく、Onにすれば表示する値が一律「00」、という異常（実際のファイルのプロパティ表示と比較して相違するため異常）です。
設定項目をOnとするのは、本件問題提起の発生条件というより前提条件です。

また、非表示ではどのような対応になるか、ご教示ありがとうございました。そこは納得しており申し分ありません。

あらためて当初の投稿の冒頭文を《》付で再掲致します。：
　FFFTPのファイル一覧に《表示される》タイムスタンプの《秒単位の値》が、一律ゼロになる問題を確認しました。
　特に《ローカルのファイルの秒単位がゼロ》になる点が、本件提起対象（問題点）です。

**Reply by sayurin** [reply]
何を求めているのかわかりません。

- 「ファイルのタイムスタンプの秒単位を表示する」が**OFF**の場合
   - 常に`0`秒として扱われます。
- 「ファイルのタイムスタンプの秒単位を表示する」が**ON**の場合
   - ローカル側 → ディスクの内容に従います。
   - リモート側 → ftpサーバーから送られてきた内容に従います。
      - https://github.com/ffftp/ffftp/discussions/288#discussioncomment-896965 で説明したように多くのftpサーバーは秒情報を送信しません。情報を受信していない場合は`0`になります。
      - メニューの「ファイル一覧をビューアで表示」を実行するとftpサーバーが送信した情報を確認できます。

----
> 特に《ローカルのファイルの秒単位がゼロ》になる点が、本件提起対象（問題点）です。

であれば、Explorerとffftpの画面キャプチャを提示してください。

**Reply by Power-Engine** [reply]
早速のご返答、そして仕様提示をありがとうございます。
求めているのは、仕様「ローカル側 → ディスクの内容に従います。」がそうなっていない、という問題提起への相談です。
問題点の画面キャプチャ、追って提示します。

なお現環境（外出先）は、問題事象を確認した環境とすべてが同一ではないのですが、再現せず正常に秒単位が表示されます。
FFFTPを実行しているローカルPCは同一です。ホストは異なります。
問題事象を確認したときのホストと次に接続できるのは、6/25（金）の予定です。
（ローカル側で完結する処理に原因があるであろう問題事象の再現性に、接続先が関係するのか疑問ですが。）

**Reply by tt4g** [reply]
> なお現環境（外出先）は、問題事象を確認した環境とすべてが同一ではないのですが、再現せず正常に秒単位が表示されます。

@Power-Engine ffftpのメンテナではない者ですが、
上記の点について気になるので愚見ですが書き込みさせていただきます。

『「秒」が表示されない端末』があったということで本ディスカッションをオープンされたものと思います。

一方で、上述のように『「秒」が表示される端末』があるということになると
事象が発生する端末でffftp以外のアプリケーションやツールであれば
「秒」が表示されていることを示した方が良いと考えます。

なぜならば、「ffftpはローカルディスクのファイル情報を表示している」と書き込みされていることから
「秒」が`0`秒になる要因が別にある場合（ファイルシステムが「秒」情報を提供していない、Windows OSの環境に依存する問題、など）に
ffftpを調査して原因を判明させるのは難しいと考えられるからです。

ソースコードを読む限り、ffftpはファイルの「最終更新時刻」をビューに表示しているようです。

https://github.com/ffftp/ffftp/blob/7409a8cb46412ea3125d024de3a2b917f6b1905a/filelist.cpp#L956-L965

事象が発生している端末で、「ffftpで「秒」が`0`になっているキャプチャ」と
「fftpで「秒」が`0`になっているファイルの最終更新日時のキャプチャ」を取得したうえで
ffftp以外では「最終更新日時」が`0`秒以外になっていれば、
ffftpに「秒」が表示できない何らかの不具合が無いか、調査の依頼が出来るのではないかと考えます。

「fftpで「秒」が`0`になっているファイルの最終更新日時のキャプチャ」については、
Windows OSでファイル右クリックで表示できるコンテキストメニューから「プロパティ(R)」に「最終更新日時」が秒単位まで表示されるので
該当のウィンドウのキャプチャで十分と思います。

調査の依頼を行う場合、「事象が発生している端末固有の設定」、「事象が発生する前後にOS設定の変更を行った」といった情報が分かれば原因を見つけやすくなると思います。

**Reply by Power-Engine** [reply]
tt4gさん、ソースコードを確認してまでの情報提供ありがとうございます。
ご教示いただいた画面キャプチャのポイント、承知しました。

1点だけ、念のため確認させてください。：
いただいた情報にある「最終更新時刻」「最終更新日時」といったプロパティに心当たりないため、あらためてエクスプローラーで探しましたが見当たりません。
単に「更新日時」ならエクスプローラーの詳細表示でも見えますが、それとは別のプロパティとして在るのでしょうか。
ちなみにこちらの環境でVerの値を出すとMicrosoft Windows [Version 10.0.19041.985]です。

冒頭の一行：
> 『「秒」が表示されない端末』があったということで本ディスカッションをオープンされたものと思います。

は誤解です。
秒が「表示されない」ということではなく、「一律ゼロ[00]である」ということを確認し、本件をオープンしました。

本件オープン時の投稿で冒頭に、
> FFFTPのファイル一覧に表示されるタイムスタンプの秒単位の値が、一律ゼロになる問題

と明記しているため、秒単位を表示するよう設定するのは当然の前提条件（＝記載するだけ冗長）と考えました。
しかしその記載を割愛したのが、誤解を与える素だったようです。：
　前提条件：　環境設定の［表示1］「ファイルのタイムスタンプの秒単位を表示する」チェックをOnにする。

**Reply by tt4g** [reply]
Windows 10をご利用のようなので、「ファイル」もしくは「ディレクトリ」を右クリックしたときに表示される
コンテキストメニューの中に存在する「プロパティ(R)」から「<ファイル名>のプロパティ」というウィンドウがひょうじされませんでしょうか。

次のキャプチャのようなメニューとウィンドウが表示されると思います。

![コンテキストメニュー](https://user-images.githubusercontent.com/45120617/122937897-4972df00-d3ad-11eb-80e0-51c447a8df37.png)

![プロパティ](https://user-images.githubusercontent.com/45120617/122937931-4d9efc80-d3ad-11eb-892f-4e032cda7175.png)


> 秒が「表示されない」ということではなく、「一律ゼロ[00]である」ということを確認し、本件をオープンしました。

ご指摘の通り「ffftpで`0`秒で表示される」が問題という認識です。

書いている途中で無意識に『秒が`0`で表記されている』を
『秒が表記されない』に変換して書いていたようです。

書き込み前の内容確認が甘かったです。申し訳ありません。

**Comment by sayurin** [comment]
ちなみに、FTPというプロトコルではファイル一覧は人間が目で見て判断するため、書式が定まっておらずバラバラになっています。

https://github.com/ffftp/ffftp/blob/v5.1/filelist.txt

には各種ftpサーバーが返すサンプルデータがあります。（先頭1文字は無視してください）
これを見るとわかりますが、ftpサーバーは割と秒情報を表示しません。更に古いファイルでは時分すら表示されません。そのため、ローカルファイルの秒情報はあまり重要ではありません。実際、ffftpで秒を扱うようになったのは2013/02のv1.98gからで、それ以前は常に`0`として処理されていました。

**Reply by Power-Engine** [reply]
FTPプロトコルの歴史情報と見解をありがとうございます。
> ローカルファイルの秒情報はあまり重要ではありません。

ローカルファイルの秒情報が一律ゼロになるという問題があっても、FTP通信ではそもそも秒情報を扱うことが少ない（⇒影響が少ない）ため問題視されにくい、ということでしょうか。

それはそうかもしれませんが、私は本件を、FTP通信の歴史的事情とは一切関係のない不具合問題、として提起しました。

何故ならローカルファイルのタイムスタンプの秒情報の取得処理は、FTPのプロトコルとは無関係にFFFTP内で実装され、FTPサーバーとは一切通信することなくローカルPC内で完結する、と考えるからです。
（一方、ホストファイルのタイムスタンプの秒情報の取得処理は、FPTサーバーとの通信が必要と考えます。）
また秒情報を取得して表示した後、その値がFTP通信でどう取り扱われるかによっては、ネットワークへの影響を危惧するからです。
もし以上の考えに間違いがあればご指摘いただけると幸いです。

ーー
貴見解「秒情報はあまり重要でない」（FTPプロトコル利用者間の慣習的制約？）によれば、〔影響状況〕に記載した私の危惧は、いわば「FTP通信への過剰な期待の裏返し」といったところですか？：
> 想定される影響として、ファイル一覧上の表示だけならともかく、ファイル・タイムスタンプの実態と異なる表示値異常が、ホストへのアップロードやホスト～ローカル間の同期処理に影響するのではないか、と危惧しています。

**Comment by sayurin** [comment, marked as answer]
反応がないので解決済みとします。もっと簡潔に書くようにしてください。今回であれば「エクスプローラーの表示とファイル日時が一致しない」とだけあれば十分に通じます。後は具体的な証拠を提示してください。
